packages/presentation/presentation_containers/interfaces/unsaved_changes/initialize_unsaved_changes.ts

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import {
+  BehaviorSubject,
+  combineLatest,
+  combineLatestWith,
+  debounceTime,
+  map,
+  Subscription,
+} from 'rxjs';
+import {
+  getInitialValuesFromComparators,
+  PublishesUnsavedChanges,
+  PublishingSubject,
+  runComparators,
+  StateComparators,
+} from '@kbn/presentation-publishing';
+import { HasSnapshottableState } from '../serialized_state';
+import { apiHasSaveNotification } from '../has_save_notification';
+
+export const COMPARATOR_SUBJECTS_DEBOUNCE = 100;
+
+export const initializeUnsavedChanges = <RuntimeState extends {} = {}>(
+  initialLastSavedState: RuntimeState,
+  parentApi: unknown,
+  comparators: StateComparators<RuntimeState>
+) => {
+  const subscriptions: Subscription[] = [];
+  const lastSavedState$ = new BehaviorSubject<RuntimeState | undefined>(initialLastSavedState);
+
+  const snapshotRuntimeState = () => {
+    const comparatorKeys = Object.keys(comparators) as Array<keyof RuntimeState>;
+    const snapshot = {} as RuntimeState;
+    comparatorKeys.forEach((key) => {
+      const comparatorSubject = comparators[key][0]; // 0th element of tuple is the subject
+      snapshot[key] = comparatorSubject.value as RuntimeState[typeof key];
+    });
+    return snapshot;
+  };
+
+  if (apiHasSaveNotification(parentApi)) {
+    subscriptions.push(
+      // any time the parent saves, the current state becomes the last saved state...
+      parentApi.saveNotification$.subscribe(() => {
+        lastSavedState$.next(snapshotRuntimeState());
+      })
+    );
+  }
+
+  const comparatorSubjects: Array<PublishingSubject<unknown>> = [];
+  const comparatorKeys: Array<keyof RuntimeState> = []; // index maps comparator subject to comparator key
+  for (const key of Object.keys(comparators) as Array<keyof RuntimeState>) {
+    const comparatorSubject = comparators[key][0]; // 0th element of tuple is the subject
+    comparatorSubjects.push(comparatorSubject as PublishingSubject<unknown>);
+    comparatorKeys.push(key);
+  }
+
+  const unsavedChanges = new BehaviorSubject<Partial<RuntimeState> | undefined>(
+    runComparators(
+      comparators,
+      comparatorKeys,
+      lastSavedState$.getValue() as RuntimeState,
+      getInitialValuesFromComparators(comparators, comparatorKeys)
+    )
+  );
+
+  subscriptions.push(
+    combineLatest(comparatorSubjects)
+      .pipe(
+        debounceTime(COMPARATOR_SUBJECTS_DEBOUNCE),
+        map((latestStates) =>
```

Comment:
Should we have `skip(1)` here since we are initializing `unsavedChanges` above? I briefly tested it both in the controls example **and** on Dashboard, and it seems to work

```suggestion
  subscriptions.push(
    combineLatest(comparatorSubjects)
      .pipe(
        skip(1),
        debounceTime(COMPARATOR_SUBJECTS_DEBOUNCE),
        map((latestStates) =>
```

Reply:
What is the reason for the skip? This code is taken from https://github.com/elastic/kibana/blob/main/src/plugins/embeddable/public/react_embeddable_system/react_embeddable_state.ts#L93. That logic was moved into this file.

Reply:
It prevents the unsaved changes from running unnecessarily on load - if you add the skip, you'll see that this subscription goes from running for every embeddable on load to only firing on the first change, which I think is ideal? And it still respects unsaved changes on load because we are initializing it above.

Reply:
right, thanks for explaining. Resolved in 80b45ac3640595723609fa1108e968e45576c1a2